transport/grpc_test.go

```diff
@@ -0,0 +1,273 @@
+				Addresses: []string{
+					"1:1:1:1:2345",
+				},
+				Tracer:   opentracing.NoopTracer{},
+				Encoding: "json",
+			},
+			wantErr: errGRPCNoCaller,
+		},
+		{
+			options: GRPCOptions{
+				Addresses: []string{
+					"1:1:1:1:2345",
+				},
+				Tracer: opentracing.NoopTracer{},
+				Caller: "example-caller",
+			},
+			wantErr: errGRPCNoEncoding,
+		},
+	}
+	for _, tt := range tests {
+		_, err := NewGRPC(tt.options)
+		if tt.wantErr != nil {
+			assert.Equal(t, tt.wantErr, err)
+		} else {
+			assert.NoError(t, err)
+		}
+	}
+}
+
+func TestGRPCSuccess(t *testing.T) {
+	doWithTestEnv(t, "example-caller", 5, []transport.Procedure{
+		newTestJSONProcedure("example", "Foo::Bar", testBar),
+	}, func(t *testing.T, testEnv *testEnv) {
+		request, err := newTestJSONRequest("example", "Foo::Bar", &testBarRequest{One: "hello"})
+		require.NoError(t, err)
+		response, err := testEnv.Transport.Call(context.Background(), request)
+		require.NoError(t, err)
+		require.NotNil(t, response)
+		testBarResponse := &testBarResponse{}
+		require.NoError(t, json.Unmarshal(response.Body, testBarResponse))
+		require.Equal(t, "hello", testBarResponse.One)
+	})
+}
+
+func TestGRPCError(t *testing.T) {
+	doWithTestEnv(t, "example-caller", 5, []transport.Procedure{
+		newTestJSONProcedure("example", "Foo::Bar", testBar),
+	}, func(t *testing.T, testEnv *testEnv) {
+		request, err := newTestJSONRequest("example", "Foo::Bar", &testBarRequest{Error: "hello"})
+		require.NoError(t, err)
+		_, err = testEnv.Transport.Call(context.Background(), request)
+		require.Equal(t, yarpcerrors.UnknownErrorf("hello"), err)
+	})
+}
+
+type testBarRequest struct {
+	One   string
+	Error string
+}
+
+type testBarResponse struct {
+	One string
+}
+
+func testBar(ctx context.Context, request *testBarRequest) (*testBarResponse, error) {
+	if request == nil {
+		return nil, nil
+	}
+	if request.Error != "" {
+		return nil, errors.New(request.Error)
+	}
+	return &testBarResponse{
+		One: request.One,
+	}, nil
+}
+
+func newTestJSONRequest(service string, method string, request interface{}) (*Request, error) {
+	body, err := json.Marshal(request)
+	if err != nil {
+		return nil, err
+	}
+	return &Request{
+		TargetService: service,
+		Method:        method,
+		Body:          body,
+	}, nil
+}
+
+func newTestJSONProcedure(service string, name string, handler interface{}) transport.Procedure {
+	procedure := yarpcjson.Procedure(name, handler)[0]
+	procedure.Service = service
+	return procedure
+}
+
+func doWithTestEnv(
+	t *testing.T,
+	caller string,
+	numInbounds int,
+	procedures []transport.Procedure,
+	f func(*testing.T, *testEnv),
+) {
+	testEnv, err := newTestEnv(caller, numInbounds, procedures)
+	require.NoError(t, err)
+	defer func() {
+		assert.NoError(t, testEnv.Close())
+	}()
+	f(t, testEnv)
+}
+
+type testEnv struct {
+	Caller         string
+	Transport      TransportCloser
+	YARPCTransport *grpc.Transport
+	YARPCInbounds  []*grpc.Inbound
+}
+
+func newTestEnv(
```

Comment:
`newGRPCTestEnv`? This scope is shared across all `yab/transport`

Reply:
+1, maybe even `runGRPCTest`

Reply:
Done